### PR TITLE
Add support to enable frame gpu capture

### PR DESCRIPTION
**Pull request by PierreCapo**
### Short description 📝

Add support to edit GPU frame capture mode on the Scheme. This will allow for instance to fix bugs when working with `MKMapView` ([stackoverflow link](https://stackoverflow.com/a/33331769))

### How to test the changes locally 🧐

- Unit tests are working properly

### Checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase.
- [x] The changes have been tested following the [documented guidelines](https://docs.tuist.io/contributors/testing-strategy/).
- [x] The PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, whenever it should be included in the “Added”, “Fixed” or “Changed” section of the CHANGELOG. Note: when included in the CHANGELOG, the title of the PR will be used as entry, please make sure it is clear and suitable.
- [ ] In case the PR introduces changes that affect users, the documentation has been updated.
- [ ] In case the PR introduces changes that affect how the cache artifact is generated starting from the `TuistGraph.Target`, the `Constants.cacheVersion` has been updated.

### CODE REVIEW

**Comment by PierreCapo** [comment]
I was having an issue while setting up the dev environment for tuist. Basically although I followed the [document guidelines](https://docs.tuist.io/contributors/get-started/) to contribute (having `generate` argument and `working directory` option) set up, it fails to execute 9 times out of 10 (sometimes it work for whatever reason).
This even happens on `main`branch, so I got hard time testing my changes.

image

Let me know also how do I edit the documentation, I tried to look at the documentation or previous PRs but I didn't really understood how I can achieve it.

Thanks in advance

File: Sources/ProjectDescription/RunActionOptions.swift [hunk 1]
```diff
@@ -24,14 +27,18 @@ public struct RunActionOptions: Equatable, Codable {
     ///
     ///     - simulatedLocation: The simulated GPS location to use when running the app.
     ///     Please note that the `.custom(gpxPath:)` case must refer to a valid GPX file in your project's resources.
+    ///
+    ///     - enableGPUFrameCaptureMode: The capture mode to use. e.g: .disabled
```
Comment (tychota) [inline Sources/ProjectDescription/RunActionOptions.swift:31]: Hi @PierreCapo, I think you can make `The capture mode to use` a bit more clear. As it is, I'm not sure I would understand it and I think it is quite an obscure options for most iOS dev.

Here is a suggestion:

```suggestion
    ///     - enableGPUFrameCaptureMode: The Metal Frame Capture mode to use. e.g: .disabled
    ///     If your target links to the Metal framework, Xcode enables GPU Frame Capture.
    ///     You can disable it to test your app in best perfomance.
```

Source https://developer.apple.com/documentation/metal/debugging_tools/enabling_frame_capture

File: Sources/ProjectDescription/RunActionOptions.swift [hunk 2]
```diff
@@ -128,3 +140,16 @@ extension RunActionOptions {
         }
     }
 }
+
+extension RunActionOptions {
+    public enum GPUFrameCaptureMode: String, Codable, Equatable {
+        case autoEnabled = "0"
+        case metal = "1"
+        case openGL = "2"
+        case disabled = "3"
```
Comment (kwridan) [inline Sources/ProjectDescription/RunActionOptions.swift:149]: There isn't a need to declare the underlying Xcode values here ... this is handled by the XcodeProj library

**Review by kwridan — COMMENTED** [review]
Thanks for this contribution @PierreCapo 🎉

It may be helpful to extend one the existing fixtures to leverage this new option as a show case / demo (e.g. https://github.com/tuist/tuist/tree/main/projects/tuist/fixtures/ios_app_with_custom_scheme)

**Comment by kwridan** [comment]
Regarding the failure in Xcode - it seems the error is indicating the project you were `tuist generate` against requires running `tuist fetch` first? Perhaps running against a fixture that doesn't have additional dependencies may help test those changes (e.g. the suggested update of the fixtures in Tuist)

**Comment by PierreCapo** [comment]
Hi @kwridan thanks a lot for the fast review! 🙏

I have been trying to adress your changes, but I have been struggling. Here is my diff (which is the last commit of my PR for debugging purpose):

```diff
diff --git a/Sources/ProjectDescription/RunActionOptions.swift b/Sources/ProjectDescription/RunActionOptions.swift
index 51de88984..f84e89696 100644
--- a/Sources/ProjectDescription/RunActionOptions.swift
+++ b/Sources/ProjectDescription/RunActionOptions.swift
@@ -1,4 +1,5 @@
 import Foundation
+import XcodeProj
 
 /// Options for the `RunAction` action
 public struct RunActionOptions: Equatable, Codable {
@@ -13,7 +14,7 @@ public struct RunActionOptions: Equatable, Codable {
     public let simulatedLocation: SimulatedLocation?
 
     /// Configure your project to work with the Metal frame debugger.
-    public let enableGPUFrameCaptureMode: GPUFrameCaptureMode
+    public let enableGPUFrameCaptureMode: XCScheme.LaunchAction.GPUFrameCaptureMode
 
     /// Creates an `RunActionOptions` instance
     ///
@@ -35,7 +36,7 @@ public struct RunActionOptions: Equatable, Codable {
         language: SchemeLanguage? = nil,
         storeKitConfigurationPath: Path? = nil,
         simulatedLocation: SimulatedLocation? = nil,
-        enableGPUFrameCaptureMode: GPUFrameCaptureMode = GPUFrameCaptureMode.default
+        enableGPUFrameCaptureMode: XCScheme.LaunchAction.GPUFrameCaptureMode = .autoEnabled
     ) {
         self.language = language
         self.storeKitConfigurationPath = storeKitConfigurationPath
@@ -64,7 +65,7 @@ public struct RunActionOptions: Equatable, Codable {
         language: SchemeLanguage? = nil,
         storeKitConfigurationPath: Path? = nil,
         simulatedLocation: SimulatedLocation? = nil,
-        enableGPUFrameCaptureMode: GPUFrameCaptureMode = GPUFrameCaptureMode.default
+        enableGPUFrameCaptureMode: XCScheme.LaunchAction.GPUFrameCaptureMode = .autoEnabled
     ) -> Self {
         self.init(
             language: language,
@@ -145,15 +146,4 @@ extension RunActionOptions {
     }
 }
 
-extension RunActionOptions {
-    public enum GPUFrameCaptureMode: String, Codable, Equatable {
-        case autoEnabled = "0"
-        case metal = "1"
-        case openGL = "2"
-        case disabled = "3"
-
-        public static var `default`: GPUFrameCaptureMode {
-            .autoEnabled
-        }
-    }
-}
+extension XCScheme.LaunchAction.GPUFrameCaptureMode: Codable {}
diff --git a/Sources/TuistGenerator/Generator/SchemeDescriptorsGenerator.swift b/Sources/TuistGenerator/Generator/SchemeDescriptorsGenerator.swift
index 92d0d667b..a4dd30353 100644
--- a/Sources/TuistGenerator/Generator/SchemeDescriptorsGenerator.swift
+++ b/Sources/TuistGenerator/Generator/SchemeDescriptorsGenerator.swift
@@ -496,7 +496,7 @@ final class SchemeDescriptorsGenerator: SchemeDescriptorsGenerating {
             )
         }
 
-        let enableGPUFrameCaptureMode: XCScheme.LaunchAction.GPUFrameCaptureMode
+       let enableGPUFrameCaptureMode: XCScheme.LaunchAction.GPUFrameCaptureMode
         if let captureMode = scheme.runAction?.options.enableGPUFrameCaptureMode {
             switch captureMode {
             case .autoEnabled:
diff --git a/Sources/TuistGraph/Models/RunActionOptions.swift b/Sources/TuistGraph/Models/RunActionOptions.swift
index f7175e845..ac69378da 100644
--- a/Sources/TuistGraph/Models/RunActionOptions.swift
+++ b/Sources/TuistGraph/Models/RunActionOptions.swift
@@ -1,5 +1,6 @@
 import Foundation
 import TSCBasic
+import XcodeProj
 
 /// Options for the `RunAction` action
 public struct RunActionOptions: Equatable, Codable {
@@ -14,7 +15,7 @@ public struct RunActionOptions: Equatable, Codable {
     public let simulatedLocation: SimulatedLocation?
 
     /// Configure your project to work with the Metal frame debugger.
-    public let enableGPUFrameCaptureMode: GPUFrameCaptureMode
+    public let enableGPUFrameCaptureMode:  XCScheme.LaunchAction.GPUFrameCaptureMode
 
     /// Creates an `RunActionOptions` instance
     ///
@@ -36,7 +37,7 @@ public struct RunActionOptions: Equatable, Codable {
         language: String? = nil,
         storeKitConfigurationPath: AbsolutePath? = nil,
         simulatedLocation: SimulatedLocation? = nil,
-        enableGPUFrameCaptureMode: GPUFrameCaptureMode = .autoEnabled
+        enableGPUFrameCaptureMode:  XCScheme.LaunchAction.GPUFrameCaptureMode = .autoEnabled
     ) {
         self.language = language
         self.storeKitConfigurationPath = storeKitConfigurationPath
@@ -90,16 +91,3 @@ extension RunActionOptions.SimulatedLocation: Equatable, Codable {
         try container.encode(identifier)
     }
 }
-
-extension RunActionOptions {
-    public enum GPUFrameCaptureMode: String, Codable, Equatable {
-        case autoEnabled = "0"
-        case metal = "1"
-        case openGL = "2"
-        case disabled = "3"
-
-        public static var `default`: GPUFrameCaptureMode {
-            .autoEnabled
-        }
-    }
-}
diff --git a/Sources/TuistLoader/Models+ManifestMappers/RunActionOptions+ManifestMapper.swift b/Sources/TuistLoader/Models+ManifestMappers/RunActionOptions+ManifestMapper.swift
index a411f055e..e276eeba4 100644
--- a/Sources/TuistLoader/Models+ManifestMappers/RunActionOptions+ManifestMapper.swift
+++ b/Sources/TuistLoader/Models+ManifestMappers/RunActionOptions+ManifestMapper.swift
@@ -2,6 +2,7 @@ import Foundation
 import ProjectDescription
 import TSCBasic
 import TuistGraph
+import XcodeProj
 
 extension TuistGraph.RunActionOptions {
     /// Maps a ProjectDescription.RunActionOptions instance into a TuistGraph.RunActionOptions instance.
@@ -15,11 +16,9 @@ extension TuistGraph.RunActionOptions {
         var language: String?
         var storeKitConfigurationPath: AbsolutePath?
         var simulatedLocation: SimulatedLocation?
-        var enableGPUFrameCaptureMode: GPUFrameCaptureMode
+        let enableGPUFrameCaptureMode = manifest.enableGPUFrameCaptureMode
 
         language = manifest.language?.identifier
-        enableGPUFrameCaptureMode = GPUFrameCaptureMode(rawValue: manifest.enableGPUFrameCaptureMode.rawValue) ??
-            GPUFrameCaptureMode.default
 
         if let path = manifest.storeKitConfigurationPath {
             storeKitConfigurationPath = try generatorPaths.resolveSchemeActionProjectPath(path)
```

Tuist compiles fine however it crashes while opening the fixture project. The stacktrace is not super clear about why it is crashing. I guess this is because it is not able to cast properly PBXGroup (as shown in the lldb debugger), and that might comes from this line which at runtime might change the reference of GPUFrameCaptureMode ? :

`extension XCScheme.LaunchAction.GPUFrameCaptureMode: Codable {}`

I guess `GPUFrameCaptureMode` should anyway be conform to `Codable` in the XcodeProj library directly no?

image

**Comment by kwridan** [comment]
Apologies for the confusion @PierreCapo 

I didn't mean for us to depend and use `XcodeProj` as a dependency within the `ProjectDescription` target - rather I meant we don't need to specifically attempt to replicate / encode the Xcode raw values at all, we can keep the enum as `: String` without any explicit values:

```swift
public enum GPUFrameCaptureMode: String, Codable, Equatable {
         case autoEnabled
         case metal
         case openGL
         case disabled
}
```

The reason is, within the `SchemeDescriptorGenerator` we already map between the enum and the underlying `XcodeProj` types - so the raw values are implementation details which we don't need to expose or use.

For a bit of background:

- `ProjectDescription`: The public Tuist manifest APIs - which Tuist users interact with when defining their projects
- `TuistGraph`: The internal models Tuist uses - most of the time it's a 1-1 with the `ProjectDescription`, but often contains additional contextual information _(e.g. paths are expanded, dependency information resolved etc...)_
- `XcodeProj`: The underlying raw xcodeproj types

https://docs.tuist.io/contributors/generation-pipeline goes in more depth into this if you're curious.

Hope this helps

**Review by tychota — APPROVED** [review]
Thanks for addressing my feedbacks @PierreCapo, looks good to me on the documentation side.

**Review by kwridan — APPROVED** [review]
Thanks @PierreCapo - this is good to merge 🚀

**Pull request merged**